Add `todo` test modifier

BarryThePenguin opened the pull request:
Seems too simple...

Fixes #563

vdemedes commented on lib/runner.js:
```diff
@@ -46,6 +46,8 @@ util.inherits(Runner, EventEmitter);
 module.exports = Runner;
 
 optionChain(chainableMethods, function (opts, title, fn) {
+	fn = fn || function () {};
```
I think having `function noop() {}` at the top of the file and using it here instead of anonymous function inline would be better.

BarryThePenguin replied:
Cool, can do. Was also wondering if `||` is enough, or should I explicitly be checking for `typeof fn === 'undefined'`?

vdemedes replied:
Nah, `||` is enough here.

vdemedes commented:
Thanks for this PR, @BarryThePenguin!

sindresorhus commented:
Can you also update the docs?

https://github.com/sindresorhus/ava#testskiptitle-body
https://github.com/sindresorhus/ava#skip-tests

sindresorhus commented:
This allows omitting test function for all tests, not just `.skip`, @vdemedes are we ok with that?

vdemedes commented:
@sindresorhus Very good point. I guess we came back to the problems discussed in #9, `test.todo()` makes more sense in that case. But for now, maybe we should restrict function-less tests only to `.skip`.

BarryThePenguin commented:
Mmm... in this context skip and todo are synonymous with each other. Maybe it's worth exploring?

BarryThePenguin commented:
Also, noting the api change, this allows for `test.skip()`. Not sure how helpful that is to people

kentcdodds commented:
> Also, noting the api change, this allows for test.skip(). Not sure how helpful that is to people

That makes sense. I wonder though if we should think about explicitly disallowing that to avoid people abusing it and confusing others.

BarryThePenguin commented:
I'd be happy to investigate the possibility of having both a `.skip()` and `.todo()` modifier.

Thoughts so far: 
- skip has optional title, requires a test function
- todo requires a test title, won't accept a test function

Would also require adding todo to the tap reporter output

kentcdodds commented:
You know, I would actually prefer to leave `skip` as it is and introduce a `todo` that only accepts a message. I would still like to have the `todo` in the output so it's a constant reminder. I think that's the value of coding the todo rather than just using a comment.

BarryThePenguin commented:
I've had a go at the `todo` modifier. I also had a go at implementing the tap output for both `todo` and `skip` directives

novemberborn commented on lib/reporters/tap.js:
```diff
@@ -27,6 +27,10 @@ TapReporter.prototype.start = function () {
 
+	var passed = test.todo ? 'not ok' : 'ok';
+	var todoDirective = test.todo ? '# TODO' : '';
+	var skipDirective = test.skipped ? '# SKIP' : '';
```
Perhaps do:

```
var directive = '';
if (test.todo) {
  directive = '# TODO';
} else if (test.skipped) {
  directive = '# SKIP';
}
```

Which simplifies the `todoDirective || skipDirective || ''` statement below and removes the unnecessary `|| ''` fall-through.

BarryThePenguin replied:
Ha, great. Will do :hankey:

novemberborn commented on lib/runner.js:
```diff
@@ -46,6 +50,12 @@ util.inherits(Runner, EventEmitter);
+		throw new TypeError('Expected a string');
+	} else if (opts.todo) {
+		fn = noop;
+	}
```
I'd prefer a single `if (opts.todo) {}`, then nest the `title` check. Took me a while to figure out `fn = noop` was used when `title` was not a function :smile:

kentcdodds replied:
In addition, I would make the check be `typeof title !== 'string'` :+1:

BarryThePenguin replied:
:+1:

BarryThePenguin replied:
FYI...

```js
var title = 'todo';
typeof typeof title !== 'string';
// === false
```

:sleeping: :zzz: :zzz:

kentcdodds replied:
lol

novemberborn commented on lib/runner.js:
```diff
@@ -46,6 +50,12 @@ util.inherits(Runner, EventEmitter);
 
 optionChain(chainableMethods, function (opts, title, fn) {
+	if (opts.todo && typeof title === 'function') {
+		throw new TypeError('Expected a string');
```
The error message could be more informative. Perhaps `TODO tests need a title` or something along those lines.

BarryThePenguin replied:
I can change it. Anyone else want to clarify before I do?

novemberborn commented on lib/runner.js:
```diff
@@ -20,6 +23,7 @@ var chainableMethods = {
 		before: {type: 'before'},
 		after: {type: 'after'},
 		skip: {skipped: true},
+		todo: {skipped: true, todo: true},
```
It might be better to not set `skipped: true` here. That would require further clauses where `skipped` is used to determine whether the test is skipped, but it also removes exceptions where you have to check for `!todo`.

@vdemedes thoughts?

BarryThePenguin replied:
Up until now I've been treating todo as a skipped test, that has failed, mainly based on the TAP spec. Kind of depends on how tightly integrated the runner implementation is with the reporter output.

BarryThePenguin replied:
Any other suggestions on what to do here?

novemberborn replied:
I guess on the one hand todo tests should be skipped, but on the other hand they should not be counted as being skipped. I'm not even sure whether they should be counted as tests now!

Not annotating the test as `skipped` requires explicit guards in the runner where it decides whether to skip a test, but also reduces `skipped && !todo` checks where doing skip counts. Bit of a mixed bag, personally I prefer the more explicit approach.

@vdemedes @sindresorhus?

vdemedes replied:
> I'm not even sure whether they should be counted as tests now!

I agree with that. But for the sake of simplicity, we can just mark them as skipped tests internally, to prevent any kind of execution.

BarryThePenguin replied:
If the todo test needs to be explicit, I'm happy to make that change

novemberborn replied:
Nah there's no clear consensus either way. It's fine as it is, just different trade-offs in different areas.

novemberborn commented on readme.md:
```diff
@@ -248,12 +248,14 @@ test.only('will be run', t => {
 ### Skip-tests
 
-Skip-tests are shown in the output as skipped but never run.
+Skip-tests are shown in the output as skipped but never run. Skip-tests can be written without specifying the test function. This can be useful for planning future tests.
```
If I'm understanding things correctly this needs to be changed to reflect `test.todo` right?

BarryThePenguin replied:
This will need to change, yes

novemberborn commented on test/reporters/tap.js:
```diff
@@ -146,3 +146,39 @@ test('results does not show skipped tests if there are none', function (t) {
+	var actualOutput = reporter.test({
+		title: 'should think about doing this',
+		passed: false,
+		todo: true
```
If we're sticking to having both `skipped` and `todo` in the metadata then you'll have to add `skipped: true` here.

kentcdodds commented:
I'm not certain I follow the results of all the changes. But I'll just say that I would recommend that `skip` require a function and the only way to avoid having to specify a function is if you use `todo`. Fewer ways to do the same thing is :+1: :D

novemberborn commented:
> But I'll just say that I would recommend that skip require a function and the only way to avoid having to specify a function is if you use `todo`. Fewer ways to do the same thing is :+1: 

Yes agreed. Would be good to make `test.skip('my test')` fail with a helpful error, pointing at `test.todo()`.

@BarryThePenguin changes are looking good generally, nice work. Would be good to clear up some of the logic and state management so it's easier to follow how these special test cases (pun very much intended) are handled.

BarryThePenguin commented:
Just went through the review.

Added Mini and Verbose reporter handling for todo + tests

vdemedes commented:
I like it, LGTM from me!

@sindresorhus @novemberborn @jamestalmage ?

sindresorhus commented on lib/runner.js:
```diff
@@ -51,6 +55,16 @@ optionChain(chainableMethods, function (opts, title, fn) {
+			throw new TypeError('Expected a string');
+		}
+	} else if (opts.skipped && typeof fn !== 'function') {
+		throw new TypeError('Expected a function. Use todo for tests without a function');
```
```
throw new TypeError('Expected a function. Use `test.todo()` for tests without a function.');
```

sindresorhus commented:
Looks pretty good. The docs needs to be updated and some inline feedback.

sindresorhus commented on lib/runner.js:
```diff
@@ -51,6 +55,16 @@ optionChain(chainableMethods, function (opts, title, fn) {
+		fn = noop;
+
+		if (typeof title !== 'string') {
+			throw new TypeError('Expected a string');
```
```
throw new TypeError('`todo` tests require a title');
```

BarryThePenguin commented:
Updated. Happy to squash/rebase to make the commit history a little more sane :+1:

sindresorhus commented:
LGTM

sindresorhus commented:
> Happy to squash/rebase to make the commit history a little more sane

:+1:

novemberborn commented on lib/reporters/tap.js:
```diff
@@ -27,6 +27,15 @@ TapReporter.prototype.start = function () {
+
+	if (test.todo) {
+		directive = '# TODO';
+	} else if (test.skipped) {
```
This should be `test.skip`, from <https://github.com/BarryThePenguin/ava/blob/bb3c972cc4df4b3df0139b5d6f738cb18da0e6e7/lib/runner.js#L99>.

novemberborn commented on lib/runner.js:
```diff
@@ -63,10 +77,12 @@ optionChain(chainableMethods, function (opts, title, fn) {
 		this.stats.testCount++;
 
-		if (result.result.metadata.skipped) {
+		if (test.metadata.skipped && !test.metadata.todo) {
```
You still need to increase the `todoCount`, as well as initialize it to `0` in the `run()` method. Then in <https://github.com/sindresorhus/ava/blob/0e1886582833986b896c31859aa61d8da404e996/api.js#L233> you need to compute the total `todoCount`. These counts are mocked in the reporter tests.

novemberborn replied:
Oh and here <https://github.com/sindresorhus/ava/blob/0e1886582833986b896c31859aa61d8da404e996/api.js#L207>.

vdemedes commented on lib/reporters/tap.js:
```diff
@@ -27,6 +27,15 @@ TapReporter.prototype.start = function () {
+	var passed = test.todo ? 'not ok' : 'ok';
+
+	if (test.todo) {
+		directive = '# TODO';
```
If I'm not mistaken, there's no such directive as `# TODO` in TAP spec. For TAP, we could treat `todo` tests just like `skip` and output them the same way.

novemberborn replied:
[There are](https://testanything.org/tap-version-13-specification.html#todo-tests).

vdemedes replied:
In that case, great! :)

kentcdodds commented:
This is awesome! Thanks everyone for your work on this. Looking forward to seeing this merged :-)

sindresorhus commented:
Still some things to iron out I think.

```js
import test from 'ava';
test.todo('foo');
test.todo('bar');
```

Gives no output with the default reporter:

```
~/dev
❯ ava          



                    
```

And with `--verbose`:

```
~/dev
❯ ava --verbose

  - foo
  - bar

  2 tests failed


                  
```

I don't think todo tests should be failing.

-

```js
import test from 'ava';
test('foo', t => t.pass());
test.todo('bar');
```

Outputs it's passing, but no mention of the `1 todo` and the test run still exit with exit code 1 even though the output indicates differently:

```
~/dev
❯ ava

  1 passed

                  
```

novemberborn commented on lib/runner.js:
```diff
@@ -63,10 +77,14 @@ optionChain(chainableMethods, function (opts, title, fn) {
 
-		if (result.result.metadata.skipped) {
+		if (test.metadata.todo) {
+			this.stats.todoCount++;
```
You need to initialize `todoCount` to `0` in the `Runner#run()` method.

BarryThePenguin replied:
Done

novemberborn commented on test/runner.js:
```diff
@@ -190,9 +190,38 @@ test('skip test', function (t) {
 	runner.run().then(function () {
 		t.is(runner.stats.testCount, 2);
 		t.is(runner.stats.passCount, 1);
+		t.is(runner.stats.failCount, 1);
```
Need to add an assertion for the `todoCount`.

novemberborn commented:
@BarryThePenguin would you mind adding a test in `test/api.js` that verifies the `passCount`, `skipCount`, `todoCount` and `failCount` values? There is no explicit test for at the moment. The reporter tests use mocked values so everything seems fine until it's verified by hand.

BarryThePenguin commented:
@sindresorhus sure, I'll see if I can work these out

I'm starting to question my reasoning for initially marking todo tests as failed. I've been following the TAP spec and I'm not sure it's quite clear.

>These tests represent a feature to be implemented or a bug to be fixed and act as something of an executable “things to do” list. They are not expected to succeed. Should a todo test point begin succeeding, the harness should report it as a bonus. This indicates that whatever you were supposed to do has been done and you should promote this to a normal test point.

This implies that todo tests fail until they are implemented. This PR currently expects todo tests to have no implementation. Does this mean we mark them as failed? Again, I'm not sure how tightly we want AVA coupled with the TAP spec.

BarryThePenguin commented:
Also, thanks everyone for providing feedback and spending the time to help out. It's been a really enjoyable experiance :hamster:

sindresorhus commented:
The TAP description doesn't explicitly say they should fail either, though, just that they shouldn't succeed like a normal test. I think we should rather enable users to fail todo tests on the [linting step](https://github.com/sindresorhus/eslint-plugin-ava/pull/32), not in AVA.

novemberborn commented:
> I think we should rather enable users to fail todo tests on the linting step, not in AVA. Todo tests should be like TODO comments, informative, but not affecting.

:+1:

BarryThePenguin commented:
Updated.

@vdemedes @novemberborn todo is no longer skipped as well. Instead, in test-collection we check for the skipped and todo modifier.

@novemberborn I've added the `test/api.js` test

@sindresorhus I'm not quite sure about the exit code being 1, when should the exit code be 1? These are the outputs for the examples you provided above

mini reporter
```

  2 todo

```

verbose reporter
```

  - foo
  - bar

  0 tests passed
  2 tests todo

```

test + todo example

```

  1 passed  1 todo

```

sindresorhus commented:
@BarryThePenguin Perfect. Exactly what I was thinking :)

LGTM

vdemedes commented:
Looks fine to me as well, thanks @BarryThePenguin!

sindresorhus commented:
Landed. Thank you for tirelessly working on this one @BarryThePenguin. It turned out really good.

![unicorn18](https://cloud.githubusercontent.com/assets/170270/13521567/26d0a690-e21c-11e5-8631-3c0af9bd6347.gif)

jfmengels commented:
Thanks @BarryThePenguin!

kentcdodds commented:
This is terrific. I love the level of thought for quality coming from this project. Thanks everyone (especially @BarryThePenguin)!
